src/lib/statistics/statistics_objects/abstract_histogram.hpp

```diff
@@ -0,0 +1,256 @@
+#pragma once
+
+#include <limits>
+#include <memory>
+#include <optional>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "histogram_domain.hpp"
+#include "statistics/statistics_objects/abstract_statistics_object.hpp"
+#include "storage/base_segment.hpp"
+#include "types.hpp"
+
+namespace opossum {
+
+using BinID = size_t;
+constexpr BinID INVALID_BIN_ID{std::numeric_limits<BinID>::max()};
+
+/**
+ * Used for HistogramBin heights and distinct counts
+ */
+using HistogramCountType = Cardinality;
+
+template <typename T>
+struct HistogramBin {
+  HistogramBin(const T& min, const T& max, const HistogramCountType height, const HistogramCountType distinct_count)
+      : min(min), max(max), height(height), distinct_count(distinct_count) {}
+
+  T min{};
+  T max{};
+  HistogramCountType height{};
+  HistogramCountType distinct_count{};
```

Comment:
In person. I'm making a list

Reply:
Still opposed to making struct members const. If C++ were const-by-default this would be a different matter, but since it isn't the const qualifier makes it look like elaborate thought went into it, and people will be afraid to remove it again. A dumb struct should look like a dumb struct. 

[This](https://www.reddit.com/r/cpp/comments/8wbeom/coding_guideline_avoid_const_member_variables/) thread provides more, maybe obscure points. 
* "[...] your type is not going to be naturally move assignable, or swappable"
* "In general, using `const` in a member variable is bad because you are enforcing an API invariant at too low a level" <- I guess this is mostly my point
* " `const` on a member variable can force your class to have a throwing move"